server/middleware/quota_manager.js

```diff
@@ -0,0 +1,131 @@
+var MiniEventEmitter = require('miniee'),
+    QuotaLimiter = require('./quota_limiter.js'),
+    Client = require('../../client/client.js'),
+    logging = require('minilog')('radar:quota_manager');
+
+var QuotaManager = function() { 
+  this._limiters = {};
+};
+
+MiniEventEmitter.mixin(QuotaManager);
+
+QuotaManager.prototype.checkLimits = function(socket, message, messageType, next) {
+  var limiter = this.getLimiter(messageType);
+
+  if (!limiter || (message.op !== 'subscribe' && message.op !== 'sync')) {
+    next();
+  } else if (limiter.isAboveLimit(socket.id)) {
+    logging.warn('#socket.message - rate_limited', message, socket.id);
+
+    socket.send({
+      op: 'err',
+      value: 'rate limited',
+      origin: message
+    });
+
+    next('limit reached');
+  } else {
+    if (messageType && messageType.policy && messageType.policy.softLimit) {
+      // Log Soft Limit, if available. 
+      softLimit = messageType.policy.softLimit;
+      if (limiter.count(socket.id) === softLimit) {
+        var client = Client.get(socket.id);
+        this._logLimits(client, softLimit, rateLimiter.count(socket.id));
+      }
+    }
+
+    next();
+  }
+};
+
+QuotaManager.prototype.updateLimits = function(socket, resource, message, messageType, next) {
+  var limiter = this.getLimiter(messageType);
+
+  if (limiter) {
+    switch(message.op) {
+      case 'sync':
+      case 'subscribe': 
+        limiter.add(socket.id, message.to);
+        break;
+      case 'unsubscribe': 
+        limiter.remove(socket.id, message.to);
+        break;
+    }
+  }
+
+  next();
+};
+
+QuotaManager.prototype.destroyByClient = function(socket, resource, messageType, next) {
+  var limiter = this.findLimiter(messageType);
+
+  if (limiter) {
+    limiter.remove(socket.id, resource.to);
+  }
+
+  next();
+};
+
+QuotaManager.prototype.destroyByResource = function(resource, messageType, next) {
+  var to = resource.to,
+      limiter = this.findLimiter(messageType);
+
+  if (limiter) {
+    limiter.removeByTo(to);
+  }
+
+  next();
+};
+
+QuotaManager.prototype.findLimiter = function(messageType) {
+  return this._limiters[messageType.name];
+};
+
+QuotaManager.prototype.getLimiter = function(messageType) {
+  var limiter = this.findLimiter(messageType);
+
+  if (!limiter && this._shouldLimit(messageType)) {
+    limiter = this._buildLimiter(messageType);
+    this._limiters[messageType.name] = limiter;
+    this.emit('rate_limiter:add', messageType.name, limiter);
```

Comment:
we probably need to document these events, so devs can get at them quickly. They are very useful.